File: jianzhioffer/firstround/jianzhi_53_1.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if not nums or (target < nums[0]) or (target > nums[-1]):
            return 0

        i, j = 0, len(nums)-1
        while i <= j:   # 先查找右边界
            mid = (i + j) // 2
            if nums[mid] <= target:
                i = mid + 1
            else:
                j = mid - 1

        if j >= 0 and nums[j] != target:   # 如果右边界的值不等于target，说明数组中没有target，直接返回0
            return 0
        right = i
        i = 0
        while i <= j:   # 查找左边界
            mid = (i + j) // 2
            if nums[mid] < target:
                i = mid + 1
            else:
                j = mid - 1
        left = j
        return right - left - 1


if __name__ == '__main__':
    arr = [5, 7, 7, 8, 8, 10]
    obj = Solution()
    num = obj.search(arr, 8)
    print(num)

File: jianzhioffer/secondround/jianzhi_53_1.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution
{
public:
    int search(vector<int>& nums, int target);
};

// //哈希表。时间复杂度O(N)，空间复杂度O(N)
// int Solution::search(vector<int>& nums, int target)
// {
//     if(nums.empty()) return 0;
//     unordered_map<int, int> map;
//     for (int i = 0; i < nums.size(); i++)
//     {
//         map[nums[i]]++;
//     }
    
//     return map[target];
// }

//二分查找。分别查找出数组中target的左边界和右边界，即可得到target的数量
//时间复杂度O(logN)，空间复杂度O(1)
int Solution::search(vector<int>& nums, int target)
{
    if(nums.empty()) return 0;
    int left = 0, right = nums.size()-1;

    while (left<=right)
    {
        int mid = (left+right)/2;
        if(nums[mid]<=target) left = mid+1;
        else right = mid-1;
    }
    int res_right = left; //target右边界的后一个位置。因为left==right时会执行最后一次循环，最后一次循环结束后，left一定比right大1，所以left指向区间右边界的后一个位置，而right指向区间的右边界
    
    if(right<0 || (right>=0 && nums[right]!=target)) return 0; //找到右边界后，判断一下数组中是否存在target，不存在则返回0

    left = 0;
    while(left<=right)
    {
        int mid = (left+right)/2;
        if(nums[mid]<target) left = mid+1;
        else right = mid-1;
    }
    int res_left = right;  //target左边界的前一个位置

    return res_right - res_left - 1;
}

int main(int argc, char const *argv[])
{
    vector<int> nums = {5,7,7,8,8,10};
    Solution obj;
    int result = obj.search(nums, 5);
    return 0;
}
